### ISSUE TITLE
Pyright does not resolve some imports that python resolves

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#4230]
**Describe the bug**
Pyright does not resolve some imports that python resolves.

**To Reproduce**
* Set up the initial state via the following shell script (run in a new directory):
```sh
# Create a fake repository
mkdir repo
echo "from foo.bar import Bar" > repo/bug.py
cat > repo/pyrightconfig.json << EOF
{
    "executionEnvironments": [
        {
            "root": "."
        }
    ],
    "venv": "venv",
    "venvPath": ".."
}
EOF
 
# Create a venv
python -m venv venv
 
# Go to the site-packages directory
cd venv/lib/python*/site-packages
 
# Create fake "foo.bar" packages (via .pth files)
mkdir -p Bar2/foo/bar
echo "Bar2" > Bar2.pth
mkdir -p Bar3/foo/bar
cat > Bar3/foo/bar/__init__.py << EOF
class Bar:
    pass
EOF
echo "Bar3" > Bar3.pth
```
* Notice that Pyright does not resolve `Bar` in `repo/bug.py`.
* Run `./venv/bin/python` and notice that `from foo.bar import Bar` works there.

**Expected behavior**
Pyright should resolve `foo.bar.Bar`, like Python does.

**Additional context**
Interestingly, the bug does NOT happen when doing any of the following things on the initial state:
* Delete the empty `Bar2/foo/bar` directory. This suggests that the bug is related to namespace packages.
* Rename `Bar3.pth` as `Bar1.pth`. This suggests that the order of discovery matters.
* Reduce the package "depth" (i.e. use `bar` everywhere instead of `foo.bar`).

**Comment by erictraut** [comment]
What you're doing here is quite unusual. You're creating two implicit namespace packages `Bar2` and `Bar3` but placing a traditional package inside one of them (`Bar3`) and leaving the other one empty. The Python loader's behavior in this case is not well documented. I would encourage you to either use implicit namespace packages or traditional packages (those that have an `__init__.py` for each submodule in the hierarchy) rather than combining the two.

That said, I've updated pyright's import resolution logic to match the apparent behavior of the Python import resolution logic in this case. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.282, which I just published. It will also be included in a future release of pylance.

**Comment by ipkiss42** [comment]
Thank you for the quick update and for all your awesome work on Pyright!

FWIW, I agree with you that this is unusual and arguably playing with fire. I found this pattern in another teams' code and made them aware of it, but I have no control over whether they change it or not.
